synapse/handlers/message.py

```diff
@@ -336,6 +337,25 @@ def create_event(self, requester, event_dict, token_id=None, txn_id=None,
             prev_events_and_hashes=prev_events_and_hashes,
         )
 
+        # In an ideal world we wouldn't need the second part of this condition. However,
+        # this behaviour isn't spec'd yet, meaning we should be able to deactivate this
+        # behaviour, and this code is evaluated each time a new m.room.aliases event is
+        # created, which includes hitting a /directory route, therefore not including it
+        # would render the similar check in synapse.handlers.directory pointless.
+        if builder.type == EventTypes.Aliases and self.require_membership_for_aliases:
+            # Ideally we'd do this check in event_auth.check(), however this function
+            # describes a spec'd algorithm and this limitation is lacking from the spec,
```

Comment:
for me, the important point is that the checks in `event_auth.check` apply to events received over federation, as well as those created locally - and since the spec (as of room v3) allows alias events to be created by users who are not in the room, we have to tolerate them.